pkg/cosign/cosign.go

```diff
@@ -145,23 +147,32 @@ func getFulcioRoots(roots []byte) (*x509.CertPool, error) {
 
 // FetchAttestations retrieves signed attestations and decodes them into in-toto statements
 // https://github.com/in-toto/attestation/blob/main/spec/README.md#statement
-func FetchAttestations(imageRef string, key string, repository string, log logr.Logger) ([]map[string]interface{}, error) {
+func FetchAttestations(imageRef string, imageVerify *v1.ImageVerification, log logr.Logger) ([]map[string]interface{}, error) {
 	ctx := context.Background()
-	var pubKey signature.Verifier
 	var err error
 
-	if strings.HasPrefix(key, "-----BEGIN PUBLIC KEY-----") {
-		pubKey, err = decodePEM([]byte(key))
-		if err != nil {
-			return nil, errors.Wrap(err, "decode pem")
+	cosignOpts := &cosign.CheckOpts{
+		ClaimVerifier: cosign.IntotoSubjectClaimVerifier,
+	}
+
+	if imageVerify.Key != "" {
+		if strings.HasPrefix(imageVerify.Key, "-----BEGIN PUBLIC KEY-----") {
+			cosignOpts.SigVerifier, err = decodePEM([]byte(imageVerify.Key))
+		} else {
+			cosignOpts.SigVerifier, err = sigs.PublicKeyFromKeyRef(ctx, imageVerify.Key)
 		}
 	} else {
-		pubKey, err = sigs.PublicKeyFromKeyRef(ctx, key)
-		if err != nil {
-			return nil, errors.Wrap(err, "loading public key")
+		cosignOpts.CertEmail = ""
+		cosignOpts.RootCerts, err = getFulcioRoots([]byte(imageVerify.Roots))
+		if err == nil {
+			cosignOpts.RekorClient, err = rekor.NewClient("https://rekor.sigstore.dev")
```

Comment:
This might be a future enhancement but we should also allow custom rekor URLs.

Reply:
Yes, we should! Logged: https://github.com/kyverno/kyverno/issues/3224.